units/commit_queue.go

```diff
@@ -0,0 +1,352 @@
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/level"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	commitQueueJobName = "commit-queue"
+)
+
+func init() {
+	registry.AddJobType(commitQueueJobName, func() amboy.Job { return makeCommitQueueJob() })
+}
+
+type commitQueueJob struct {
+	job.Base `bson:"job_base" json:"job_base" yaml:"job_base"`
+	QueueID  string `bson:"queue_id" json:"queue_id" yaml:"queue_id"`
+}
+
+func makeCommitQueueJob() *commitQueueJob {
+	job := &commitQueueJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    commitQueueJobName,
+				Version: 0,
+			},
+		},
+	}
+	job.SetDependency(dependency.NewAlways())
+
+	return job
+}
+
+func NewCommitQueueJob(queueID string, id string) amboy.Job {
+	job := makeCommitQueueJob()
+	job.QueueID = queueID
+	job.SetID(fmt.Sprintf("%s:%s_%s", commitQueueJobName, queueID, id))
+
+	return job
+}
+
+func (j *commitQueueJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	projectRef, err := model.FindOneProjectRef(j.QueueID)
+	if err != nil {
+		j.AddError(errors.Wrapf(err, "can't find project for queue id %s", j.QueueID))
+		return
+	}
+
+	if !projectRef.CommitQEnabled {
+		return
+	}
+
+	cq, err := commitqueue.FindOneId(j.QueueID)
+	if err != nil {
+		j.AddError(errors.Wrapf(err, "can't find commit queue for id %s", j.QueueID))
+	}
+	nextItem := cq.Next()
+	nextItemInt, err := strconv.Atoi(nextItem)
+	if err != nil {
+		j.AddError(errors.Wrapf(err, "can't parse next item \"%s\" as int", nextItem))
+		return
+	}
+
+	alreadyProcessed, err := versionExists(nextItemInt)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "can't query for versions for this item"))
+	}
+	if alreadyProcessed {
+		return
+	}
+
+	githubToken, err := getGitHubToken()
+	if err != nil {
+		j.AddError(errors.Wrap(err, "can't get github token"))
+	}
+
+	pr, err := thirdparty.GetGithubPullRequest(ctx, githubToken, projectRef.Owner, projectRef.Repo, nextItemInt)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "can't get PR from GitHub"))
+		return
+	}
+
+	if validatePR(pr) != nil {
+		j.AddError(errors.Wrap(err, "invalid PR"))
+		return
+	}
+
+	// GitHub hasn't yet tested if the PR is mergeable.
+	// Check back later
+	// See: https://developer.github.com/v3/pulls/#response-1
+	if pr.Mergeable == nil {
+		return
+	}
+
+	if !*pr.Mergeable {
+		err = sendCommitQueueGithubStatus(pr, message.GithubStateFailure, "PR not mergeable", "")
+		if err != nil {
+			j.AddError(errors.Wrap(err, "can't send github status"))
+		}
+		return
+	}
+
+	v, err := makeVersion(ctx, githubToken, projectRef, pr)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "can't make version"))
+		return
+	}
+
+	err = sendCommitQueueGithubStatus(pr, message.GithubStatePending, "preparing to test merge", v.Id)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "can't send github status"))
+	}
+
+	err = subscribeMerge(projectRef, pr, v.Id)
```

Comment:
This may be something for a different ticket, but if the merge attempt fails won't that block the queue? Do we need to add some handling of that here?

Reply:
well, the sender is supposed to pop the queue whether or not the merge is successful. However, if the sender never gets called for some reason or errors/returns before getting to pop the queue then like you're saying we would block indefinitely on the head of the queue.

Maybe when we check if a patch has already been created to make sure we aren't making a duplicate we can check the time of the existing patch and if it's over some threshold try again?

Reply:
I think that's reasonable, but we'll want some way to time out an entire merge attempt. It's definitely not for this PR though, but keep that in mind when you work on the merging part

Reply:
The merge part should already be good (from EVG-5580).